students/rachel_s/session04/mailroom_with_dicts.py

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
+
+import csv
+import os
+import datetime
+
+# Import donor data
+
+donor_info = {}
+
+with open('mailroom/donors.csv') as csv_file:
+    csv_reader = csv.reader(csv_file, delimiter=',')
+    for line in csv_reader:
+        if not line[0].lower() in donor_info:
+            donor_info[line[0].lower()] = [int(line[1])]
+        else:
+            donor_info[line[0].lower()].append(int(line[1]))
+
+
+def user_menu():
+    menu = {1: send_thank_you, 2: create_report, 3: all_donor_letters, 4: quit}
+    response = input("Select an option:\n  1  Send a thank you to a single donor\n  2  Create a report\n  3  Send letters to all donors\n  4  Quit\n\n> ")
+    if int(response) in menu.keys():
+        menu.get(int(response))()
+    else:
+        response = input("Please select 1, 2, 3, or 4.\n> ")
+
+
+# Send a thank you
+def send_thank_you():
+    donor_name = input("What is the donor's full name?\n> ").lower()
+    if donor_name == 'list':
+        for i in donor_info.keys():
+            print(i.title())
+        user_menu()
```

Comment:
Uhh... You are using undesired recursion for the main menu loop. It's best not to use recursion for that purpose. It's confusing, like it's not obvious how the loop works, when some functions call `user_menu()` at the end, but other functions don't. I thought you used the desired while loop in your mailroom part 1, and I'm wondering why you changed that structure... Please see [this](https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom_tutorial.html#controlling-main-program-flow) regarding the recommended main menu loop control structure.